core/domain/html_validation_service.py

```diff
@@ -820,54 +818,156 @@ def regenerate_image_filename_using_dimensions(filename, height, width):
+        # So we check for the existense of this attribute.
+        if (not collapsible_component.has_attr('content-with-value') or
+                collapsible_component['content-with-value'] == ''):
+            continue
+        # Create a new soup with the content-with-value html string.
+        content_with_html_string = unescape_html(
+            collapsible_component['content-with-value'])
+        content_soup = bs4.BeautifulSoup(
+            content_with_html_string, 'html.parser')
+        # Modify the filenames for the image tags in the content soup.
+        _modify_image_filename(exploration_fs, content_soup)
+        # content_soup contains the updated filepath. This is copied
+        # into the content-with-value attribute of the original soup.
+        collapsible_component['content-with-value'] = (
+            escape_html(python_utils.UNICODE(content_soup).replace('\'', '')))
+
+    # To add dimensions to images inside the tab component.
+    for tab_component in soup.findAll(
+            name='oppia-noninteractive-tabs'):
+        # The tab_contents-with-value attribute contains the html with image
+        # tags. So we check for the existense of this attribute.
+        if (not tab_component.has_attr('tab_contents-with-value') or
+                tab_component['tab_contents-with-value'] == ''):
+            continue
+        # Create a new soup with the tab_contents-with-value html string.
+        content_with_html_string = unescape_html(
+            tab_component['tab_contents-with-value'])
+        content_soup = bs4.BeautifulSoup(
+            content_with_html_string, 'html.parser')
+        # Modify the filenames for the image tags in the content soup.
+        _modify_image_filename(exploration_fs, content_soup)
+        # content_soup contains the updated filepath. This is copied
+        # into the tab_contents-with-value attribute of the original soup.
+        tab_component['tab_contents-with-value'] = (
+            escape_html(python_utils.UNICODE(content_soup)
+                        .replace('\'', '')))
+    return python_utils.UNICODE(soup).replace('<br/>', '<br>')
+
+
+def _modify_image_filename(exploration_fs, soup):
+    """Modifies filenames of images. This is a helper method for
+    add_dims_to_img_in_complex_rte.
+
+    Args:
+        exploration_fs: GcsFileSystem. File system of the Exploration entity.
+        soup: bs4.BeautifulSoup. The html soup whose image file is
+            to be renamed.
+    """
+    for image in soup.findAll(name='oppia-noninteractive-image'):
+        if (not image.has_attr('filepath-with-value') or
+                image['filepath-with-value'] == '' or
+                image['filepath-with-value'] == '\\"""\\"'):
+            image.decompose()
+            continue
+
+        try:
+            filename = (
+                json.loads(unescape_html(image['filepath-with-value']
+                                         .replace('\\"', ''))))
```

Comment:
This needs some explanation.

I'm also a little confused about this, it seems unusually manual. Are you sure you're applying the escaping and the json decomposition in the right order? If so, then where do the extra \\" come from?